### PR TITLE
sql: make now() and friends return timestamptz

### PR DESCRIPTION
**Pull request by dt**
also allow comparing timestamp to timestamp tz

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9618)

### CODE REVIEW

File: sql/testdata/datetime [hunk 1]
```diff
@@ -167,7 +167,7 @@ BEGIN TRANSACTION
 statement ok
 CREATE TABLE kv (
   k CHAR PRIMARY KEY,
-  v TIMESTAMP
+  v TIMESTAMP WITH TIME ZONE
```
Comment (vivekmenezes) [inline sql/testdata/datetime]: Can you take out the explicit casting on line 130

**Review by vivekmenezes — COMMENTED** [review]
Can you add some more tests for all the operations you've added. Thanks

**Comment by knz** [comment]
this breaks `insert into foo(ts_col) values (now())`

(ಥ_ಥ)

**Comment by dt** [comment]
well, we _could_ make `timestamp` == `timstatmp with time zone` and make `without time zone` the strange one, but that'd be a departure from sql spec

**Comment by knz** [comment]
I think there's a in-between situations we could live with:
- have `age`, `x - y` etc operate on `timestamptz` only like you're doing here
- enable compare between ts and tstz, like you're doing here
- add an overload of now(), xxx_timestamp() etc so both a version returning ts and tstz exist side by side.

I think this covers the use cases:

- `now() - age(x, y)`: age selects tstz, minus selects its variant that operates on tstz only, and this forces to use the now() overload that produces tstz
- `insert into ts_col values (now())` still works

What do you think?

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-:-KSrJDKE1AW0wlGNw7Uh:b-z2ej1c)*

**Comment by vivekmenezes** [comment]
#9627

**Comment by dt** [comment]
Over the weekend I played with the variant of this that adds `now` overloads for both types, to make existing usages of `now` as a timestamp keep working. It created quite a mess of casts and ascriptions all over anything using `now`, convincing me that forcing manual disambiguation is not a great solution, and making me think #9637 might be the best compromise until we decide where to go with 1 vs 2 timestamp types.

**Comment by dt** [comment]
@knz adding #9637 to typing only complicates any overloads that opt-in to using it, and we can confine that to just `now` and friends until we come up with something more general (implicit conversions, possibly only for on-disk col vs expr, etc) so I think it's currently the most promising solution I have for landing this fix.

**Comment by dt** [comment]
@knz @nvanbenschoten I've rebased this on top of #9637 and propose this as an intermediate state until we figure out if we can kill one of our timestamp types (without losing ORM compat). PTAL.

**Comment by nvanbenschoten** [comment]
Interestingly enough, this is the current blocker for GORM unit tests. Specifically, their unit tests issue the query:

```
CREATE TABLE animals (
  counter BIGSERIAL, 
  name TEXT DEFAULT 'galeone', 
  "from" TEXT, 
  age TIMESTAMP WITH TIME ZONE DEFAULT "current_timestamp"(),
  created_at TIMESTAMP WITH TIME ZONE, 
  updated_at TIMESTAMP WITH TIME ZONE, 
  PRIMARY KEY (counter)
)
```
(note the age default value)

**Comment by dt** [comment]
@nvanbenschoten I believe this change should resolve that? That is, _if_ you and @knz are okay with the addition to `overloads.go`

**Comment by knz** [comment]
Reviewed 1 of 4 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/pgwire_test.go, line 464 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrAuJoI0eZeeMCg9C:-KTLrAuKUfxlZ72RMgoZ:b-y9rjhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/pgwire_test.go#L464)):*
> ```Go
> 			),
> 		},
> 		"INSERT INTO d.ts VALUES(STATEMENT_TIMESTAMP()::timestamp, $1) RETURNING b": {
> ```

I thought there is now automatic overload selection. Why is it not used here?

---

*[sql/parser/builtins.go, line 610 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrTKsYUJwZLmVkPkg:-KTLrTKsYUJwZLmVkPkh:ba7wgp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/parser/builtins.go#L610)):*
> ```Go
> 		Builtin{
> 			Types:      ArgTypes{},
> 			ReturnType: TypeTimestampTZ,
> ```

I think we want overloads with a preference for this one and the two following, too.

---

*[sql/testdata/alter_table, line 424 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrs0DCw9lBFf-vX66:-KTLrs0DCw9lBFf-vX67:b31zs6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/testdata/alter_table#L424)):*
> ```
> # Add a column with a default statement_timestamp()
> statement ok
> ALTER TABLE add_default ADD COLUMN e TIMESTAMPTZ DEFAULT statement_timestamp()
> ```

This shouldn't matter, right? (see my previous comment about overloading this ts too)
 I'd suggest testing both, i.e. one column of each type but with the same statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618)*

**Comment by dt** [comment]
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[sql/parser/builtins.go, line 610 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrTKsYUJwZLmVkPkg:-KTPC0WOZ-RRwa6BBZRp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/parser/builtins.go#L610)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think we want overloads with a preference for this one and the two following, too.
</details>
Done.

---

*[sql/testdata/datetime, line 170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KSqdgHQNrgvGHg6kh1M-r1-170:-KTPD8ueqwGGFbP_3JKp:b-q9zarg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83446ad7f47711350452e7111d7685041babce0/sql/testdata/datetime#L170)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Can you take out the explicit casting on line 130
</details>
we still need the explicit case from to timestamp or to timestamptz, since there's no implicit coercion to `TypeDate`

---

*[sql/pgwire_test.go, line 464 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrAuJoI0eZeeMCg9C:-KTPC-0mdcv0fRbl38T4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/pgwire_test.go#L464)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I thought there is now automatic overload selection. Why is it not used here?
</details>
Done.

---

*[sql/testdata/alter_table, line 424 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTLrs0DCw9lBFf-vX66:-KTPCYXIL2C8Aw9_8PDy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d01bd9e1611366693c7a638c889f36f2b927f7/sql/testdata/alter_table#L424)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This shouldn't matter, right? (see my previous comment about overloading this ts too)
>  I'd suggest testing both, i.e. one column of each type but with the same statement.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618)*

**Comment by knz** [comment]
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 614 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTPTednMiK3uRWLEAuN:-KTPTedoqCBfZ9skLTY8:b-6qngk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c63d48be6693dd75d3082553f5e98561b7e72/sql/parser/builtins.go#L614)):*
> ```Go
> 			impure:            true,
> 			fn: func(ctx *EvalContext, args DTuple) (Datum, error) {
> 				return ctx.GetStmtTimestamp(), nil
> ```

The better code would be to have GetStmtTimestamp etc return a time.Time instead of DTimestamp(,TZ), and call the MAke function here instead.

---

*[sql/parser/eval.go, line 1476 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTPTnCLe5LKnwQmhxgK:-KTPTnCMkMUEGNU4gGJ6:bdiquss) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c63d48be6693dd75d3082553f5e98561b7e72/sql/parser/eval.go#L1476)):*
> ```Go
> // GetStmtTimestampNoTz retrieves the current statement timestamp as per
> // the evaluation context. The timestamp is guaranteed to be nonzero.
> func (ctx *EvalContext) GetStmtTimestampNoTz() *DTimestamp {
> ```

See previous comment, use just 1 function and call Make in the caller.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618)*

**Comment by dt** [comment]
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 614 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTPTednMiK3uRWLEAuN:-KTPztK-I3x9uJB69DNe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c63d48be6693dd75d3082553f5e98561b7e72/sql/parser/builtins.go#L614)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> The better code would be to have GetStmtTimestamp etc return a time.Time instead of DTimestamp(,TZ), and call the MAke function here instead.
</details>
Done.

---

*[sql/parser/eval.go, line 1476 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTPTnCLe5LKnwQmhxgK:-KTPzrYTMJBT1em_g4rQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c63d48be6693dd75d3082553f5e98561b7e72/sql/parser/eval.go#L1476)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> See previous comment, use just 1 function and call Make in the caller.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-:-KTQ0h1qfRsAW0AL3yPJ:bnfp4nl)*

**Comment by dt** [comment]
AFAIK, there's still a pending objection from @nvanbenschoten to adding the overload preference, but my current opinion is that this is probably our best way forward until we figure out if we can either a) add some timestamp<->tz coercions or b) move towards having only one timestamp type, to improve the usability of timestamps.

**Comment by nvanbenschoten** [comment]
No objections from my end at this point once the remaining issues in the other PR are resolved and an issue is filed. Thanks for the help with this @dt.

**Comment by dt** [comment]
Added tests to `overloads_test.go` to assert that the `preferredOverload` is picked iff there is only one as suggested by @nvanbenschoten.

**Comment by dt** [comment]
@nvanbenschoten I think #9627 serves to keep track of the desire to unify to a single type, so I just added a note to it saying that in the event it doesn't succeed, we should instead explore implicit coercions to kill the overload picking.

**Comment by nvanbenschoten** [comment]
:lgtm: but consider not using actual Builtins in `overload_test.go`, as our type system should be independent from our set of Builtin functions.

---

Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/overload_test.go, line 85 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-KTolHrx_PRaH0-QpEVG:-KTolHrx_PRaH0-QpEVH:b-cnj4k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/01d47cc71f1fc63b56ac444287a400bb0044c9fc/sql/parser/overload_test.go#L85)):*
> ```Go
> 	binaryIntDateFn := makeTestOverload(TypeDate, TypeInt, TypeDate)
> 
> 	now := make([]overloadImpl, 0, len(Builtins["now"]))
> ```

Ideally these would use `makeTestOverload` (or `makePreferredTestOverload`) so that they're not tied to actual Builtins. I'd hate for our set of NOW() builtins to change in the future and for this test to break because of the dependency here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-:-KTolwd9Wgdneb3QsyVG:b5cek84)*

**Comment by dt** [comment]
@nvanbenschoten I'd initially avoided adding it to the `overloadImpl` interface and cluttering up the other impl's, instead sniffing for Builtin in particular in overload resolution, but that forces its use in the test. I'll switch to add it to the interface and fix the test.

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9618#-:-KTp7sSi1i_13ECbYq5K:bnfp4nl)*

**Pull request merged**